### PR TITLE
[EventHubs] add fixed backoff retry mode

### PR DESCRIPTION
**Pull request by swathipil**
fixes: #8670

### CODE REVIEW

File: sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py [hunk 1]
```diff
@@ -72,6 +72,8 @@ class EventHubConsumerClient(ClientBase):
      The failed internal partition consumer will be closed (`on_partition_close` will be called if provided) and
      new internal partition consumer will be created (`on_partition_initialize` will be called if provided) to resume
      receiving.
+    :keyword retry_mode: Fixed or exponential delay between attempts, default is exponential.
```
Comment (swathipil) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py:75]: @yunhaoling according to the[ issue/azure-core](https://github.com/Azure/azure-sdk-for-python/issues/8670), shouldn't we have the `retry_backoff_factor` and `retry_backoff_max` here as well? we don't have them defined in any of the docstring in sync/async producer/consumer clients.

File: sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py [hunk 2]
```diff
@@ -154,6 +155,13 @@ def _build_uri(address, entity):
     address += "/" + str(entity)
     return address
 
+def _get_backoff_time(retry_mode, backoff_factor, retried_times):
+    if retry_mode == RetryMode.Fixed:
```
Comment (swathipil) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py:159]: @yunhaoling in azure-core, they do a check on `retried_times`  and ["ignore redirects"](https://github.com/Azure/azure-sdk-for-python/blob/35d5436cf58a207613cc2bfd1fa796c67f92af1e/sdk/core/azure-core/azure/core/pipeline/policies/_retry.py#L112). It looks like we don't do this check. Is this b/c we're using amqp here? Or b/c we are checking this already elsewhere?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py [hunk 3]
```diff
@@ -154,6 +155,13 @@ def _build_uri(address, entity):
...
+        backoff_value = backoff_factor
+    else:
+        backoff_value = backoff_factor * (2 ** retried_times)
+    return backoff_value
```
Comment (swathipil) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py:163]: @yunhaoling we also just log if the backoff_value is more than the max (on line 351), but don't actually retry. However, in azure-core, it looks like they still retry with the [max_backoff if this backoff_value is greater ](https://github.com/Azure/azure-sdk-for-python/blob/35d5436cf58a207613cc2bfd1fa796c67f92af1e/sdk/core/azure-core/azure/core/pipeline/policies/_retry.py#L121)than allowed. Why don't we do this here?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from typing import Optional, Dict, Any
+
+from enum import Enum
+
+# INCLUDE?: _LOGGER = logging.getLogger(__name__)
```
Comment (swathipil) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py:9]: I don't think the logger is needed here, right? 
azure-core only has it b/c they also have the RetryPolicy methods in this file? https://github.com/Azure/azure-sdk-for-python/blob/6bcbe7f6897ebe140cfa5531dbc89a08d1a2f140/sdk/core/azure-core/azure/core/pipeline/policies/_retry.py#L50

File: sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py (hunk 2 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py:159]: We're checking [reties_time](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py#L540) here.

from core code itself, I think it does backoff starting from the second time when an error (consecutive error) is met.
as far as I know we're not doing redirect in eh/sb, so I think we could ignore "ignore redirect" for now.

File: sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py (hunk 3 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_client_base.py:163]: good catch, I think this is a bug, we should keep retrying as the core does

File: sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py (hunk 1 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py:75]: yep, those two are missing and we need to add them

File: sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py [hunk 5]
```diff
@@ -72,6 +72,15 @@ class EventHubConsumerClient(ClientBase):
      The failed internal partition consumer will be closed (`on_partition_close` will be called if provided) and
      new internal partition consumer will be created (`on_partition_initialize` will be called if provided) to resume
      receiving.
+    :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+     (most errors are resolved immediately by a second try without a delay).
+     In fixed mode, retry policy will always sleep for {backoff factor}.
+     In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+     seconds. If the backoff_factor is 0.1, then the retry will sleep
+     for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+    :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py:81]: Python supports float-type timeout as well.
```suggestion
    :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py [hunk 6]
```diff
@@ -43,6 +43,15 @@ class EventHubProducerClient(ClientBase):
     :keyword str user_agent: If specified, this will be added in front of the user agent string.
     :keyword int retry_total: The total number of attempts to redo a failed operation when an error occurs. Default
      value is 3.
+    :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+     (most errors are resolved immediately by a second try without a delay).
+     In fixed mode, retry policy will always sleep for {backoff factor}.
+     In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+     seconds. If the backoff_factor is 0.1, then the retry will sleep
+     for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+    :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py:52]:
```suggestion
    :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py [hunk 7]
```diff
@@ -178,6 +187,15 @@ def from_connection_string(cls, conn_str, **kwargs):
         :keyword str user_agent: If specified, this will be added in front of the user agent string.
         :keyword int retry_total: The total number of attempts to redo a failed operation when an error occurs.
          Default value is 3.
+        :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+         (most errors are resolved immediately by a second try without a delay).
+         In fixed mode, retry policy will always sleep for {backoff factor}.
+         In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+         seconds. If the backoff_factor is 0.1, then the retry will sleep
+         for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+        :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py:196]:
```suggestion
        :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from typing import Optional, Dict, Any
+
+from enum import Enum
+
+class RetryMode(str, Enum):
+    Exponential = 'exponential'
+    Fixed = 'fixed'
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py:11]: as per the guideline: DO use UPPERCASE names for enum names. https://azure.github.io/azure-sdk/python_design.html#enumerations
```suggestion
    EXPONENTIAL = 'exponential'
    FIXED = 'fixed'
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_client_async.py [hunk 9]
```diff
@@ -79,6 +79,15 @@ class EventHubConsumerClient(ClientBaseAsync):
      The failed internal partition consumer will be closed (`on_partition_close` will be called if provided) and
      new internal partition consumer will be created (`on_partition_initialize` will be called if provided) to resume
      receiving.
+    :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+     (most errors are resolved immediately by a second try without a delay).
+     In fixed mode, retry policy will always sleep for {backoff factor}.
+     In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+     seconds. If the backoff_factor is 0.1, then the retry will sleep
+     for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+    :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_client_async.py:88]:
```suggestion
    :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_client_async.py [hunk 10]
```diff
@@ -239,6 +248,15 @@ def from_connection_string(
          information. The failed internal partition consumer will be closed (`on_partition_close` will be called
          if provided) and new internal partition consumer will be created (`on_partition_initialize` will be called if
          provided) to resume receiving.
+        :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+         (most errors are resolved immediately by a second try without a delay).
+         In fixed mode, retry policy will always sleep for {backoff factor}.
+         In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+         seconds. If the backoff_factor is 0.1, then the retry will sleep
+         for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+        :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_client_async.py:257]:
```suggestion
        :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/aio/_producer_client_async.py [hunk 11]
```diff
@@ -46,6 +46,15 @@ class EventHubProducerClient(ClientBaseAsync):
     :keyword str user_agent: If specified, this will be added in front of the user agent string.
     :keyword int retry_total: The total number of attempts to redo a failed operation when an error occurs. Default
      value is 3.
+    :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+     (most errors are resolved immediately by a second try without a delay).
+     In fixed mode, retry policy will always sleep for {backoff factor}.
+     In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+     seconds. If the backoff_factor is 0.1, then the retry will sleep
+     for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+    :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/aio/_producer_client_async.py:55]:
```suggestion
    :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/aio/_producer_client_async.py [hunk 12]
```diff
@@ -196,6 +205,15 @@ def from_connection_string(
         :keyword str user_agent: If specified, this will be added in front of the user agent string.
         :keyword int retry_total: The total number of attempts to redo a failed operation when an error occurs.
          Default value is 3.
+        :keyword float retry_backoff_factor: A backoff factor to apply between attempts after the second try
+         (most errors are resolved immediately by a second try without a delay).
+         In fixed mode, retry policy will always sleep for {backoff factor}.
+         In 'exponential' mode, retry policy will sleep for: `{backoff factor} * (2 ** ({number of total retries} - 1))`
+         seconds. If the backoff_factor is 0.1, then the retry will sleep
+         for [0.0s, 0.2s, 0.4s, ...] between retries. The default value is 0.8.
+        :keyword int retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/aio/_producer_client_async.py:214]:
```suggestion
        :keyword float retry_backoff_max: The maximum back off time. Default value is 120 seconds (2 minutes).
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/__init__.py [hunk 13]
```diff
@@ -33,5 +34,6 @@
     "LoadBalancingStrategy",
     "PartitionContext",
     "parse_connection_string",
-    "EventHubConnectionStringProperties"
+    "EventHubConnectionStringProperties",
+    "RetryMode"
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/__init__.py:38]: There're API changes, remember to get @annatisch to take a look at the API view before CC day :)

**Review by yunhaoling — COMMENTED** [review]
it looks great, thanks!
left some minor comments

File: sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py (hunk 8 above)
Comment (swathipil) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_retry.py:11]: 🤦 thanks!

**Review by yunhaoling — APPROVED** [review]
I'm approving this PR, but I would recommend letting @annatisch review the API changes first before merging it to avoid rolling back.

**Pull request merged**